### PR TITLE
CMaizeProject

### PR DESCRIPTION
**Pull request by zachcran**
**Is this pull request associated with an issue(s)?**
No, but it is related to the [CMaize Workflow image](https://raw.githubusercontent.com/CMakePP/CMaize/master/docs/src/developer/cmaize_workflow.png) and [Overall Design image](https://raw.githubusercontent.com/CMakePP/CMaize/master/docs/src/developer/overall_design.png).

**Description**
The CMaize project is intended to capture information about the workflow to build and package the project. This PR implements the `CMaizeProject` class from the [Overall Design image](https://raw.githubusercontent.com/CMakePP/CMaize/master/docs/src/developer/overall_design.png) and corresponding user API from the [CMaize Workflow image](https://raw.githubusercontent.com/CMakePP/CMaize/master/docs/src/developer/cmaize_workflow.png). The existing user API functions for adding executable, library, and testing targets will also be modified to add those targets to the `CMaizeProject`.

**TODOs**
- [x] Create `CMaizeProject` class.
- [x] Create `cmaize_project()` user function.
- [x] Add targets created in other user API functions to the `CMaizeProject`.
- [x] Unit test `CMaizeProject`.
- [x] Integration testing.

### CODE REVIEW

**Review by ryanmrichard — COMMENTED** [review]
Combined with what we talked about yesterday I think this is on the right track.

**Comment by zachcran** [comment]
@ryanmrichard I believe this is ready for review.

There are a few issues that don't affect us much at the moment, but could come up later. I am not sure how to fix them right now, so maybe we could just note them as issues for now. Let me know if you have any suggestions to solve them.

1. Right now, the `CMaizeProject` `languages` attribute is not being set to anything when a `CMaizeProject` is created automatically through an `include(get_cmaize)` call, so it is essentially a useless attribute. As far as I know, I can't recover what was passed to the `LANGUAGES <languages>` keyword of the prior `project()` call. CMake does not provide a variable that lists which enabled languages are associated with a project either. I thought about defaulting `languages` to the currently enabled languages, but that is not always going to be accurate since the list of currently enabled languages is all languages enabled by all projects. Defaulting to C and CXX, like CMake does when a new `project()` call is made without a language specified, also won't always be accurate (imagine a prior `project(test_proj LANGUAGES FORTRAN)` call, now our assumption is wrong when `CMaizeProject` is auto-created).

2. Allowing `CMaizeProject(CTOR` to create a CMake `project()` itself, as opposed to calling `project()` prior to a  `CMaizeProject(CTOR` call with the same project name, will create the CMake project, but once the CTOR is done, the project appears to leave scope and all of its corresponding project variables are cleared.

    a. The only exception I have found is when there has not been a `project()` call in the `CMakeLists.txt` file yet, in which case you will get a `No project() command is present. The top-level CMakeLists.txt file must contain a literal, direct call to the project() command.` warning, but the project created in the `CMaizeProject(CTOR` will take over as the top-level project.

**Comment by ryanmrichard** [comment]
On 1. I think it suffices to leverage the `cmaize_cxx_library` etc. calls to fill in the language field. Basically the first time you see `cmaize_cxx_library` (or executable) you tag the project as needing C++.

I'm not sure I'm following 2. I thought we were going to require the user to call `project`?

File: cmake/cmaize/project/project.cmake [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+        CMaizeProject(SET "${self}" name "${_ctor_name}")
+
+        CMaizeProject(SET "${self}" languages "${_ctor_LANGUAGES}")
+
+        # If the project name is different than the current project, assume
+        # a new CMake project needs to be created
+        if(NOT "${_ctor_name}" STREQUAL "${PROJECT_NAME}")
+            project("${_ctor_name}" ${ARGN})
+        endif()
```
Comment (ryanmrichard) [inline cmake/cmaize/project/project.cmake:118]: Maybe throw an error instead?

Comment (zachcran) [reply cmake/cmaize/project/project.cmake:118]: Ah, I didn't realize calling `project` was going to be required. This can definitely be an error instead.

**Comment by zachcran** [comment]
> On 1. I think it suffices to leverage the `cmaize_cxx_library` etc. calls to fill in the language field. Basically the first time you see `cmaize_cxx_library` (or executable) you tag the project as needing C++.

Good idea! I'll work that in.

**Review by ryanmrichard — DISMISSED** [review]

**Comment by zachcran** [comment]
@ryanmrichard r2g

I tested the languages being added to a project during `cmaize_add_library` and `cmaize_add_target` calls in an example repo and it appears to be working. 

Regarding the unit tests failing: Although the unit tests were working on my local machine with CMake 3.20.2, none of the versions we are testing worked until I explicitly included `cmaize/targets/build_target`. No clue why it worked on my system, because I don't think it should have.

**Review by ryanmrichard — APPROVED** [review]

**Pull request merged**